### PR TITLE
feat(drawer): add drawer outline model for non-rectangular drawers

### PR DESCRIPTION
**Pull request by andymai**
## Summary

First PR toward #2528 (non-rectangular baseplates for L/⊓-shaped drawers). Instead of the issue's rectilinear-mask framing, the design models the **drawer outline** as a first-class concept — a single closed CCW loop of line segments + circular arcs (DXF vertex+bulge encoding, drawer-local mm) — so diagonal corner cuts and curved fronts are covered by the same mechanism as rectilinear notches. Cell painting, bin-footprint tracing, corner cuts, and a pen editor become authoring surfaces that all write this one field.

This PR is the pure-TS foundation: types + shared geometry/model modules + tests. Nothing user-visible; no behavior change for layouts without an outline.

## Changes

- `core/types.ts`: `DrawerOutline` / `OutlineVertex` / `CornerCut(Params)` + optional `Drawer.outline`. Vertices array typed mutable for Immer draft compatibility (the `CellMask.cells` precedent) with an immutability contract.
- `shared/utils/drawerOutlineGeometry.ts`: brepjs-free analytic engine — arc math (`bulge = tan(sweep/4)`, positive = CCW sweep bowing right of travel), memoized flattening (≤0.05mm chord error), `classifyRect` → inside/outside/partial with a 0.05mm boundary inset so grid-aligned rectilinear outlines classify abutting cells as fully covered, `insideAreaFraction` (sliver rules), `isFootprintInsideOutline` (placement test). One shared classifier for main thread AND worker so pocket selection, hatching, and split planning can never drift.
- `shared/utils/drawerOutline.ts`: validation (closed/simple/CCW/in-bounds incl. arc bellies/min one grid cell/≤256 vertices), quantization (0.01mm) + bbox snapping, translate/rotate180, FNV-1a geometry hash for cache keys, `resizeDrawerOutline` (arc-aware half-plane crop on shrink; single-weld strip splice on grow; resets to full rectangle when the result would be degenerate, disconnected, or would trap a notch as an interior hole), `normalizeDrawerOutline` read-side ingress guard (crops stale outlines, drops invalid ones).
- `shared/utils/drawerOutlineCells.ts`: memoized outside-cell set for grid hatching/fills, honoring fractional drawer edges; invariant: hatched ⇔ placement fails (same predicate).

## Follow-ups (per the multi-PR plan)

Generation (outline prism ∩ socket grid via the existing corner-rounding intersect slot), split planner support, CQRS command + server validation, layout-tab gating, then the authoring UI behind a labs flag.

## Testing

- 53 unit tests across the three modules: arc conventions pinned (incl. non-semicircle endpoint round-trip), boundary-on-gridline classification, crop-through-arc validity, grow-weld hole detection, hash jitter stability, fractional-edge cell sets, ingress normalization.
- `pnpm run quality` clean (typecheck, lint, knip).

---
## Summary by cubic
Introduces a first-class `Drawer.outline` model (CCW loop of lines + arcs) to support non-rectangular drawers (L/⊓, diagonal, curved) with a single shared classifier. Foundation for Linear #2528; no UI or behavior changes for rectangular layouts.

- **New Features**
  - Added `DrawerOutline`, `OutlineVertex`, `CornerCut(Params)`, and optional `Drawer.outline` in `core/types.ts`.
  - Implemented `shared/utils/drawerOutlineGeometry.ts`: bulge-based arc math, flattening (≤0.05mm), `classifyRect`, `insideAreaFraction`, `isFootprintInsideOutline`, shared by main thread and worker.
  - Added `shared/utils/drawerOutline.ts`: validation (closed/simple/CCW/in-bounds/area), quantization (0.01mm) and bbox snapping, translate/rotate180, geometry-only FNV-1a hash, `resizeDrawerOutline` (crop/grow with arc-aware logic), and `normalizeDrawerOutline`.
  - Added `shared/utils/drawerOutlineCells.ts`: memoized outside-cell set for hatching/fills, honors fractional edges; invariant aligns with placement checks.
  - 53 unit tests pin conventions and edge cases; `pnpm run quality` passes.

- **Bug Fixes**
  - `isFullRectangle` now uses a topology check (segments hug boundary lines) so small corner chamfers are preserved instead of being dropped by an area tolerance.
  - Increased arc flattening cap to 256 steps to keep the ≤0.05mm chord-error guarantee for large-radius arcs; clarified docs for the chord-approximate self-intersection check and `classifyRect` containment semantics.

<sup>Written for commit fa0fb1086c76d267855afd07d1e7c1970f04266a. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/andymai/gridfinity-layout-tool/pull/2530?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/shared/utils/drawerOutline.ts [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+}
+
+function isFullRectangle(outline: DrawerOutline, widthMm: number, depthMm: number): boolean {
+  for (const v of outline.vertices) {
+    const onX = Math.abs(v.x) <= OUTLINE_SNAP_EPS || Math.abs(v.x - widthMm) <= OUTLINE_SNAP_EPS;
+    const onY = Math.abs(v.y) <= OUTLINE_SNAP_EPS || Math.abs(v.y - depthMm) <= OUTLINE_SNAP_EPS;
+    if (!onX && !onY) return false;
+  }
+  const area = polylineSignedArea(flattenOutline(outline));
+  return area >= widthMm * depthMm - OUTLINE_SNAP_EPS * 2 * (widthMm + depthMm);
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/shared/utils/drawerOutline.ts]: Fixed in fa0fb1086 — replaced the area-tolerance heuristic with a topology test (every segment straight and hugging one of the four boundary lines), so a corner-cutting edge with endpoints on two different lines is never treated as a rectangle. Added a regression test: 5mm chamfer on a 10-unit drawer survives normalizeDrawerOutline.

File: src/shared/utils/drawerOutlineGeometry.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+ */
+export const BOUNDARY_EPS = 0.05;
+
+/** Max chord deviation (mm) when flattening arcs. Far below print tolerance. */
+export const ARC_FLATTEN_TOLERANCE = 0.05;
+
+/** Bulges smaller than this are treated as straight segments. */
+export const BULGE_EPS = 1e-9;
+
+const MAX_ARC_STEPS = 64;
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/shared/utils/drawerOutlineGeometry.ts]: Fixed in fa0fb1086 — raised MAX_ARC_STEPS to 256 with a comment deriving the bound: with |bulge| ≤ 1 inside a 50-unit drawer the largest valid radius (~1485mm semicircle) needs ~192 steps, so 256 never truncates a valid outline's accuracy.

File: src/shared/utils/drawerOutline.ts [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+    if (
+      p.x < -OUTLINE_SNAP_EPS ||
+      p.x > widthMm + OUTLINE_SNAP_EPS ||
+      p.y < -OUTLINE_SNAP_EPS ||
+      p.y > depthMm + OUTLINE_SNAP_EPS
+    ) {
+      return err('out_of_bounds', 'outline leaves the drawer extent');
+    }
+  }
+  if (isSelfIntersecting(pts)) {
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/shared/utils/drawerOutline.ts:190]: Documented in fa0fb1086. With the arc-step cap fixed, a crossing missed by the chord approximation is a graze shallower than ~0.1mm — below print resolution and handled by the BREP boolean — so exact arc-arc intersection isn't warranted here. The doc comment now states the approximation and its bound.

File: src/shared/utils/drawerOutlineGeometry.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+  }
+  return t0 <= t1;
+}
+
+export type RegionClass = 'inside' | 'outside' | 'partial';
+
+/**
+ * Classify an axis-aligned rect (drawer-local mm) against the outline.
+ * The rect is inset by {@link BOUNDARY_EPS} first, so boundary-coincident
+ * geometry counts as covered. 'partial' = the outline's boundary passes
```
[the remark this answers is not in the document]
Comment (andymai) [reply src/shared/utils/drawerOutlineGeometry.ts]: Fixed in fa0fb1086 — docstring now covers the containment case ('partial' = any part of the boundary lies within the inset rect, including fully contained).
